### PR TITLE
Rom app improve fomat rom parameters

### PR DESCRIPTION
**Pull request by Rbravo555**
**📝 Description**
This PR seeks to move the ROM information (namely the **nodal_modes** and the **elements_and_weights**) currently stored inside the RomParameters.json into 4 (or 5 if Petrov-Galerkin is used) numpy arrays: 
- RightBasisMatrix.npy
- LeftBasisMatrix.npy (if Petrov Galerkin is used)
- NodeIds.npy
- ElementsVector.npy
- WeightsMatrix.npy

**🆕 Changelog**
- In **calculate_rom_basis_output_process** store ModesMatrix.npy and NodeIds.npy to disk rather than appending them to the RomParameters.json
- In **hrom_training_utility** store WeightsMatrix.npy and ElementsVector.npy rather than appending them to the RomParameters
- In **rom_analysis** take the ROM data from the numpy arrays instead of the RomParameters.json

As dicussed a number of numpy arrays are stored to disk, rather than a single one that might be less transparent. The branch is fully compatible with Galerkin, LSPG and Petrov-Galerkin projections. Moreover, we have kept the option of using JSON for the ROM information if the user so decides

### CODE REVIEW

**Comment by rubenzorrilla** [comment]
At lack of further discussion, I think I like more your current proposal (separate `npy` files for each dataset). I think it eases the parsing when compared to the unique `npy` file.

File: applications/RomApplication/python_scripts/hrom_training_utility.py [hunk 1]
```diff
@@ -82,9 +82,10 @@ def CreateHRomModelParts(self):
         hrom_main_model_part = aux_model.CreateModelPart(model_part_name)
 
         # Get the weights and fill the HROM computing model part
-        with open('RomParameters.json','r') as f:
-            rom_parameters = KratosMultiphysics.Parameters(f.read())
-        KratosROM.RomAuxiliaryUtilities.SetHRomComputingModelPart(rom_parameters["elements_and_weights"], computing_model_part, hrom_main_model_part)
+        KratosROM.RomAuxiliaryUtilities.SetHRomComputingModelPart(
+            KratosMultiphysics.Parameters(json.JSONEncoder().encode(self.__CreateDictionaryWithRomElementsAndWeights())),
```
Comment (RiccardoRossi) [inline applications/RomApplication/python_scripts/hrom_training_utility.py]: i am not clear about what you are doing here. could u expand a bit?

File: applications/RomApplication/python_scripts/hrom_training_utility.py [hunk 2]
```diff
@@ -192,19 +178,60 @@ def __AppendHRomWeightsToRomParameters(self):
...
+    def __AddSelectedConditionsWithZeroWeights(self, original_weights, original_conditions, conditions_to_add, number_of_elements):
+
+        added_conditions_ids = np.array(conditions_to_add)+number_of_elements
+        updated_conditions = np.r_[original_conditions, added_conditions_ids]
+        updated_weights = np.r_[original_weights, np.zeros(added_conditions_ids.shape)]
+
+        return updated_weights, updated_conditions
+
+
+    def __CreateDictionaryWithRomElementsAndWeights(self, w = None, z=None, n_elements = None):
```
Comment (RiccardoRossi) [inline applications/RomApplication/python_scripts/hrom_training_utility.py]: does it make sense to still call this "CreateDictionary"?

File: applications/RomApplication/python_scripts/hrom_training_utility.py [hunk 3]
```diff
@@ -192,19 +178,60 @@ def __AppendHRomWeightsToRomParameters(self):
...
+        if type(z)==np.int64 or type(z)==np.int32:
+            # Only one element found !
+            if z <= n_elements-1:
+                hrom_weights["Elements"][int(z)] = float(w)
+            else:
+                hrom_weights["Conditions"][int(z)-n_elements] = float(w)
+        else:
+            # Many elements found
+            for j in range (0,len(z)):
+                if z[j] <=  n_elements -1:
```
Comment (RiccardoRossi) [inline applications/RomApplication/python_scripts/hrom_training_utility.py]: i think that here you could do this by vectorized ops using numpy.

the thing that worries me a bit is if we are sure that this is not the part that takes long

File: applications/RomApplication/python_scripts/rom_analysis.py [hunk 4]
```diff
@@ -100,13 +101,20 @@ def ModifyAfterSolverInitialize(self):
...
-                for key,value in zip(hrom_weights_elements.keys(), hrom_weights_elements.values()):
-                    computing_model_part.GetElement(int(key)+1).SetValue(KratosROM.HROM_WEIGHT, value.GetDouble()) #FIXME: FIX THE +1
+                WeightsMatrix = np.load("WeightsMatrix.npy") #FIXME: FIX THE +1
+                ElementsVector = np.load("ElementsVector.npy") #FIXME: FIX THE +1
+                #FIXME for a HROM using the Hyper-reduced model part without unselected elements, the number of elements originally present in the model part should be provided!
+                #TODO Decide whether storing an extra ingeter "OriginalNumberOfElements.npy" or saving all the HROM info as a numpy structure(less transparent to new users)
+                OriginalNumberOfElements = computing_model_part.NumberOfElements()
+                for i in range(WeightsMatrix.shape[0]):
+                    if ElementsVector[i] < OriginalNumberOfElements:
+                        computing_model_part.GetElement(int( ElementsVector[i])+1).SetValue(KratosROM.HROM_WEIGHT, WeightsMatrix[i]  )
```
Comment (RiccardoRossi) [inline applications/RomApplication/python_scripts/rom_analysis.py]: if you want efficiency this needs to be in c++...

**Comment by roigcarlo** [comment]
@RiccardoRossi, @Rbravo555 I am starting to put together the 9.3 release. If this is needed, please Add it to the milestone.

File: applications/RomApplication/python_scripts/hrom_training_utility.py [hunk 5]
```diff
@@ -206,19 +195,67 @@ def __AppendHRomWeightsToRomParameters(self):
...
+        if self.hrom_output_format=="numpy":
+            np.save('WeightsMatrix.npy',w)
+            np.save('ElementsVector.npy',z)
 
-        with open('RomParameters.json','w') as f:
-            json.dump(updated_rom_parameters, f, indent = 4)
+        elif self.hrom_output_format=="json":
+            with open('RomParameters.json','r') as f:
+                updated_rom_parameters = json.load(f)
+                updated_rom_parameters["elements_and_weights"] = hrom_weights #TODO: Rename elements_and_weights to hrom_weights
```
Comment (NicolasSR) [inline applications/RomApplication/python_scripts/hrom_training_utility.py:214]: In the json format case, you update the rom parameters locally but you're not saving them again as a json anymore. Is that alright?

**Comment by Rbravo555** [comment]
The tests have been updated!

**Comment by Rbravo555** [comment]
Either this PR is merged first and the RomManager (#10763) is adapted to include numpy, or the RomManager is merged first and this PR is adapted to use it. It would be great to have both before 03/03/2023

File: applications/RomApplication/python_scripts/calculate_rom_basis_output_process.py [hunk 6]
```diff
@@ -58,10 +58,9 @@ def __init__(self, model, settings):
 
         # Set the ROM basis output settings
         self.rom_basis_output_format = settings["rom_basis_output_format"].GetString()
-        if not self.rom_basis_output_format == "json":
-            err_msg = "Provided \'rom_basis_output_format\' is {}. Available options are \'json\'.".format(self.rom_basis_output_format)
+        if not self.rom_basis_output_format == "json" and not self.rom_basis_output_format == "numpy" :
```
Comment (rubenzorrilla) [inline applications/RomApplication/python_scripts/calculate_rom_basis_output_process.py]:
```suggestion
        rom_basis_output_available_formats = ["json", "numpy"]
        if self.rom_basis_output_format not in rom_basis_output_available_formats:
```
Just in case we add more in the future.

File: applications/RomApplication/python_scripts/calculate_rom_basis_output_process.py [hunk 7]
```diff
@@ -136,6 +135,9 @@ def ExecuteFinalize(self):
 
         # Set a NumPy array with the snapshots data
         n_nodes = self.model_part.NumberOfNodes()
+        n_elements = self.model_part.NumberOfElements()
+        rom_basis_dict["hrom_settings"]["original_number_of_elements"] = n_elements
+        rom_basis_dict["hrom_settings"]["hrom_format"] = self.rom_basis_output_format
```
Comment (rubenzorrilla) [inline applications/RomApplication/python_scripts/calculate_rom_basis_output_process.py:142]:
```suggestion
        rom_basis_dict["hrom_settings"]["output_format"] = self.rom_basis_output_format
```
I think `hrom` is redundant in this case. Besides, I also think that `output_format` is more self-descriptive.

File: applications/RomApplication/python_scripts/calculate_rom_basis_output_process.py [hunk 8]
```diff
@@ -151,20 +153,27 @@ def ExecuteFinalize(self):
...
+        elif self.rom_basis_output_format == "numpy":
+            # Storing modes in Numpy format
+            numpy.save('RightBasisMatrix.npy', u)
+            numpy.save('NodeIds.npy',  numpy.arange(1,((u.shape[0]+1)/n_nodal_unknowns), 1, dtype=int)   )
         else:
             err_msg = "Unsupported output format {}.".format(self.rom_basis_output_format)
             raise Exception(err_msg)
 
+        # Creating the RomParameters.json containing or not the modes depending on "self.rom_basis_output_format"
+        output_filename = "RomParameters.json"
```
Comment (rubenzorrilla) [inline applications/RomApplication/python_scripts/calculate_rom_basis_output_process.py]: Are we 100% sure we want to hardcode this? I think it might be handy to keep the optional naming (OFC defaulting to `RomParameters.json`).

File: applications/RomApplication/python_scripts/hrom_training_utility.py [hunk 9]
```diff
@@ -41,6 +41,7 @@ def __init__(self, solver, custom_settings):
         self.rom_settings = custom_settings["rom_settings"]
         self.hrom_visualization_model_part = settings["create_hrom_visualization_model_part"].GetBool()
         self.projection_strategy = settings["projection_strategy"].GetString()
+        self.hrom_output_format = settings["hrom_format"].GetString()
```
Comment (rubenzorrilla) [inline applications/RomApplication/python_scripts/hrom_training_utility.py:44]:
```suggestion
        self.hrom_output_format = settings["output_format"].GetString()
```
Same comment.

File: applications/RomApplication/python_scripts/hrom_training_utility.py [hunk 10]
```diff
@@ -132,6 +134,8 @@ def CreateHRomModelParts(self):
     @classmethod
     def __GetHRomTrainingDefaultSettings(cls):
         default_settings = KratosMultiphysics.Parameters("""{
+            "original_number_of_elements": 10,
```
Comment (rubenzorrilla) [inline applications/RomApplication/python_scripts/hrom_training_utility.py]:
```suggestion
            "original_number_of_elements": 0,
```
I'd default this to zero to then check that a valid value is provided, in other words, if after the defaults check the `original_number_of_elements` is still 0 we throw an error.

File: applications/RomApplication/python_scripts/hrom_training_utility.py [hunk 11]
```diff
@@ -132,6 +134,8 @@ def CreateHRomModelParts(self):
...
+            "hrom_format": "numpy",
```
Comment (rubenzorrilla) [inline applications/RomApplication/python_scripts/hrom_training_utility.py:141]:
```suggestion
            "output_format": "numpy",
```

File: applications/RomApplication/python_scripts/hrom_training_utility.py [hunk 12]
```diff
@@ -206,19 +195,67 @@ def __AppendHRomWeightsToRomParameters(self):
...
-        with open('RomParameters.json','w') as f:
-            json.dump(updated_rom_parameters, f, indent = 4)
+        elif self.hrom_output_format=="json":
+            with open('RomParameters.json','r') as f:
+                updated_rom_parameters = json.load(f)
+                updated_rom_parameters["elements_and_weights"] = hrom_weights #TODO: Rename elements_and_weights to hrom_weights
 
         if self.echo_level > 0 : KratosMultiphysics.Logger.PrintInfo("HRomTrainingUtility","\'RomParameters.json\' file updated with HROM weights.")
+
+
```
Comment (rubenzorrilla) [inline applications/RomApplication/python_scripts/hrom_training_utility.py]:
```suggestion
```
Remove useless lines.

File: applications/RomApplication/python_scripts/hrom_training_utility.py [hunk 13]
```diff
@@ -206,19 +195,67 @@ def __AppendHRomWeightsToRomParameters(self):
...
+    def __AddSelectedElementsWithZeroWeights(self, original_weights,original_elements, elements_to_add):
+
+        added_elements_ids = np.array(elements_to_add)
+        updated_elements = np.r_[original_elements, added_elements_ids]
+        updated_weights = np.r_[original_weights, np.zeros(added_elements_ids.shape)]
+
+        return updated_weights, updated_elements
+
+
```
Comment (rubenzorrilla) [inline applications/RomApplication/python_scripts/hrom_training_utility.py]:
```suggestion
```

File: applications/RomApplication/python_scripts/hrom_training_utility.py [hunk 14]
```diff
@@ -206,19 +195,67 @@ def __AppendHRomWeightsToRomParameters(self):
...
+    def __AddSelectedConditionsWithZeroWeights(self, original_weights, original_conditions, conditions_to_add, number_of_elements):
+
+        added_conditions_ids = np.array(conditions_to_add)+number_of_elements
+        updated_conditions = np.r_[original_conditions, added_conditions_ids]
+        updated_weights = np.r_[original_weights, np.zeros(added_conditions_ids.shape)]
+
+        return updated_weights, updated_conditions
+
```
Comment (rubenzorrilla) [inline applications/RomApplication/python_scripts/hrom_training_utility.py:235]:
```suggestion
```

File: applications/RomApplication/python_scripts/hrom_training_utility.py (hunk 2 above)
Comment (rubenzorrilla) [reply applications/RomApplication/python_scripts/hrom_training_utility.py]: I think it does, it returns a dictionary.

File: applications/RomApplication/python_scripts/hrom_training_utility.py [hunk 15]
```diff
@@ -206,19 +195,67 @@ def __AppendHRomWeightsToRomParameters(self):
...
+
+    def __CreateDictionaryWithRomElementsAndWeights(self, w = None, z=None, n_elements = None):
+
+        if w is None:
+            w = np.load('WeightsMatrix.npy')
+        if z is None:
+            z = np.load('ElementsVector.npy')
```
Comment (rubenzorrilla) [inline applications/RomApplication/python_scripts/hrom_training_utility.py]: As far as I understand, in here you are assuming the new numpy output formatting. Would the json-based one still work with this?

File: applications/RomApplication/python_scripts/hrom_training_utility.py [hunk 16]
```diff
@@ -206,19 +195,67 @@ def __AppendHRomWeightsToRomParameters(self):
...
+
+        if type(z)==np.int64 or type(z)==np.int32:
+            # Only one element found !
+            if z <= n_elements-1:
+                hrom_weights["Elements"][int(z)] = float(w)
+            else:
+                hrom_weights["Conditions"][int(z)-n_elements] = float(w)
+        else:
+            # Many elements found
+            for j in range (0,len(z)):
```
Comment (rubenzorrilla) [inline applications/RomApplication/python_scripts/hrom_training_utility.py]:
```suggestion
            for j in range (len(z)):
```

File: applications/RomApplication/python_scripts/rom_analysis.py [hunk 17]
```diff
@@ -110,26 +111,47 @@ def ModifyAfterSolverInitialize(self):
...
-                        aux[j,i] = nodal_modes[node_id][j][i].GetDouble()
-                node.SetValue(KratosROM.ROM_BASIS, aux)
-            
-            # Set the left nodal ROM basis if it is different than the right nodal ROM basis (i.e. Petrov-Galerkin)
-            if (self.solving_strategy == "petrov_galerkin"):
-                petrov_galerkin_nodal_modes = self.rom_parameters["petrov_galerkin_nodal_modes"]
-                petrov_galerkin_nodal_dofs = len(self.project_parameters["solver_settings"]["rom_settings"]["nodal_unknowns"].GetStringArray())
-                aux = KratosMultiphysics.Matrix(petrov_galerkin_nodal_dofs, self.petrov_galerkin_rom_dofs)
+            # Set the right nodal ROM basis
+            if self.rom_parameters["rom_format"].GetString() == "json":
```
Comment (rubenzorrilla) [inline applications/RomApplication/python_scripts/rom_analysis.py:115]:
```suggestion
            if self.rom_parameters["output_format"].GetString() == "json":
```
Same comment.

Now I realize that you might have done it this way (`rom_format` and `hrom_format`) to allow for different formats for the modes and the weights. If so, I'd make a unique format for both for the sake of simplicity.

File: applications/RomApplication/python_scripts/rom_analysis.py [hunk 18]
```diff
@@ -110,26 +111,47 @@ def ModifyAfterSolverInitialize(self):
...
+                        node_id = str(node.Id)
+                        for j in range(petrov_galerkin_nodal_dofs):
+                            for i in range(self.petrov_galerkin_rom_dofs):
+                                aux[j,i] = petrov_galerkin_nodal_modes[node_id][j][i].GetDouble()
+                        node.SetValue(KratosROM.ROM_LEFT_BASIS, aux)
+
+            elif self.rom_parameters["rom_format"].GetString() == "numpy":
+                #### Set the nodal ROM basis ####
+                Nodes = np.load("NodeIds.npy")
+                RightModes = np.load("RightBasisMatrix.npy")
```
Comment (rubenzorrilla) [inline applications/RomApplication/python_scripts/rom_analysis.py]: Maybe not for this PR, but I think that we could consider to make these filenames user-definable as it might be handy in some situations.

Comment (rubenzorrilla) [inline applications/RomApplication/python_scripts/rom_analysis.py]:
```suggestion
                nodes = np.load("NodeIds.npy")
                right_modes = np.load("RightBasisMatrix.npy")
```
Please stick to our [style guide](https://github.com/KratosMultiphysics/Kratos/wiki/Style-Guide). In this case, note that a local variable is not written in `snake_case` as it should be but in `CamelCase`.

File: applications/RomApplication/python_scripts/rom_analysis.py [hunk 19]
```diff
@@ -110,26 +111,47 @@ def ModifyAfterSolverInitialize(self):
...
+                if RightModes.ndim ==1: #check if matrix contains a single mode (a 1D numpy array)
+                    RightModes.reshape(-1,1)
+                RightModes = RightModes[:,:rom_dofs]
+                if (self.solving_strategy == "petrov_galerkin"):
+                    petrov_galerkin_rom_dofs = self.project_parameters["solver_settings"]["rom_settings"]["petrov_galerkin_number_of_rom_dofs"].GetInt()
+                    LeftModes = np.load("LeftBasisMatrix.npy")
```
Comment (rubenzorrilla) [inline applications/RomApplication/python_scripts/rom_analysis.py]:
```suggestion
                    left_modes = np.load("LeftBasisMatrix.npy")
```
Same style guide comment.

File: applications/RomApplication/python_scripts/rom_analysis.py [hunk 20]
```diff
@@ -138,14 +160,28 @@ def ModifyAfterSolverInitialize(self):
...
-                    computing_model_part.GetElement(int(key)+1).SetValue(KratosROM.HROM_WEIGHT, value.GetDouble()) #FIXME: FIX THE +1
-
-                hrom_weights_condtions = self.rom_parameters["elements_and_weights"]["Conditions"]
-                for key,value in zip(hrom_weights_condtions.keys(), hrom_weights_condtions.values()):
-                    computing_model_part.GetCondition(int(key)+1).SetValue(KratosROM.HROM_WEIGHT, value.GetDouble()) #FIXME: FIX THE +1
+                if self.rom_parameters["hrom_settings"]["hrom_format"].GetString() == "json":
+                    # Set the HROM weights in elements and conditions
+                    hrom_weights_elements = self.rom_parameters["elements_and_weights"]["Elements"]
+                    for key,value in zip(hrom_weights_elements.keys(), hrom_weights_elements.values()):
+                        print('value being imposed to element ',key , '  is:', value.GetDouble())
```
Comment (rubenzorrilla) [inline applications/RomApplication/python_scripts/rom_analysis.py]: Isn't this a debugging leftover?
```suggestion
```

File: applications/RomApplication/python_scripts/rom_analysis.py [hunk 21]
```diff
@@ -138,14 +160,28 @@ def ModifyAfterSolverInitialize(self):
...
+                        computing_model_part.GetElement(int(key)+1).SetValue(KratosROM.HROM_WEIGHT, value.GetDouble()) #FIXME: FIX THE +1
+                    hrom_weights_condtions = self.rom_parameters["elements_and_weights"]["Conditions"]
+                    for key,value in zip(hrom_weights_condtions.keys(), hrom_weights_condtions.values()):
+                        print('value being imposed to condition ',key , ' is:', value.GetDouble())
+
```
Comment (rubenzorrilla) [inline applications/RomApplication/python_scripts/rom_analysis.py]:
```suggestion
```
Same.

File: applications/RomApplication/python_scripts/rom_analysis.py [hunk 22]
```diff
@@ -138,14 +160,28 @@ def ModifyAfterSolverInitialize(self):
...
+                        computing_model_part.GetCondition(int(key)+1).SetValue(KratosROM.HROM_WEIGHT, value.GetDouble()) #FIXME: FIX THE +1
+                elif self.rom_parameters["hrom_settings"]["hrom_format"].GetString() == "numpy":
+                    # Set the HROM weights in elements and conditions
+                    WeightsMatrix = np.load("WeightsMatrix.npy")
+                    ElementsVector = np.load("ElementsVector.npy")
+                    OriginalNumberOfElements = self.rom_parameters["hrom_settings"]["original_number_of_elements"].GetInt()
```
Comment (rubenzorrilla) [inline applications/RomApplication/python_scripts/rom_analysis.py]:
```suggestion
                    weights_matrix = np.load("WeightsMatrix.npy")
                    elements_vector = np.load("ElementsVector.npy")
                    original_number_of_elements = self.rom_parameters["hrom_settings"]["original_number_of_elements"].GetInt()
```

Comment (rubenzorrilla) [reply applications/RomApplication/python_scripts/rom_analysis.py]: Same style guide comment.

File: applications/RomApplication/tests/test_calculate_rom_basis_output_process.py [hunk 23]
```diff
@@ -33,7 +33,6 @@ def testCalculateRomBasisOutputProcess(self):
             "snapshots_interval": 1.0,
             "nodal_unknowns": ["TEMPERATURE"],
             "rom_basis_output_format": "json",
-            "rom_basis_output_name": "RomParameters",
```
Comment (rubenzorrilla) [inline applications/RomApplication/tests/test_calculate_rom_basis_output_process.py]: Same comment. I think we should keep this user-definable.

File: applications/RomApplication/tests/test_compressible_potiential_rom.py [hunk 24]
```diff
@@ -1,5 +1,5 @@
 import os
-
+import numpy as np
```
Comment (rubenzorrilla) [inline applications/RomApplication/tests/test_compressible_potiential_rom.py:2]: Why is this required if not used in this file?

**Review by rubenzorrilla — COMMENTED** [review]
Some minor comments. Aside of these, it looks good to me.

File: applications/RomApplication/python_scripts/rom_analysis.py (hunk 20 above)
Comment (Rbravo555) [reply applications/RomApplication/python_scripts/rom_analysis.py]: It was a leftover, it's been cleaned

File: applications/RomApplication/python_scripts/hrom_training_utility.py (hunk 15 above)
Comment (Rbravo555) [reply applications/RomApplication/python_scripts/hrom_training_utility.py]: JSON format will still work, and all the ROM info will be stored in the RomParameters.json if the flag is set for "json" rather than "numpy"

File: applications/RomApplication/tests/test_compressible_potiential_rom.py (hunk 24 above)
Comment (Rbravo555) [reply applications/RomApplication/tests/test_compressible_potiential_rom.py:2]: For that particular test it was required but not being imported. The failing test did not trigger any alarm because the RomApplication does not apper in the Continuous Integration any longer

Comment (rubenzorrilla) [reply applications/RomApplication/tests/test_compressible_potiential_rom.py:2]: OK, but didn't we add the `RomApplication` to the CI at some point?

Comment (Rbravo555) [reply applications/RomApplication/tests/test_compressible_potiential_rom.py:2]: We just spoke with @roigcarlo and found out that the RomApplication is indeed still in the CI. It just so happens that the "nightly" test suite is empty. We only have the tests in the "small" one. We shall add it in a future breve PR

File: applications/RomApplication/python_scripts/calculate_rom_basis_output_process.py [hunk 25]
```diff
@@ -58,10 +58,10 @@ def __init__(self, model, settings):
 
         # Set the ROM basis output settings
         self.rom_basis_output_format = settings["rom_basis_output_format"].GetString()
-        if not self.rom_basis_output_format == "json":
-            err_msg = "Provided \'rom_basis_output_format\' is {}. Available options are \'json\'.".format(self.rom_basis_output_format)
+        rom_basis_output_available_formats = ["json", "numpy"]
+        if self.rom_basis_output_format not in rom_basis_output_available_formats:
+            err_msg = "Provided \'rom_basis_output_format\' is {}. Available options are \'json\' and \'numpy\'.".format(self.rom_basis_output_format)
             raise Exception(err_msg)
-        self.rom_basis_output_name = settings["rom_basis_output_name"].GetString()
```
Comment (SADPR) [inline applications/RomApplication/python_scripts/calculate_rom_basis_output_process.py:64]: In my opinion, it would be beneficial to retain the choice of "rom_basis_output_name". For example, in the context of CoSimulation for fluid-thermal coupling, it is necessary to have separate files named RomParameters_fluid.json and RomParameters_thermal.json. Therefore, I propose allowing the ability to pass the name of the "RomParameters" file to the ``RomAnalysis`` (this can be done in a future PR).

File: applications/RomApplication/python_scripts/calculate_rom_basis_output_process.py [hunk 26]
```diff
@@ -151,20 +152,27 @@ def ExecuteFinalize(self):
...
+            # Storing modes in JSON format
+            i = 0
+            for node in self.model_part.Nodes:
+                rom_basis_dict["nodal_modes"][node.Id] = u[i:i+n_nodal_unknowns].tolist()
+                i += n_nodal_unknowns
+
+        elif self.rom_basis_output_format == "numpy":
+            # Storing modes in Numpy format
+            numpy.save('RightBasisMatrix.npy', u)
+            numpy.save('NodeIds.npy',  numpy.arange(1,((u.shape[0]+1)/n_nodal_unknowns), 1, dtype=int)   )
```
Comment (SADPR) [inline applications/RomApplication/python_scripts/calculate_rom_basis_output_process.py:169]: I am curious about whether this would still work if there is a floating node within the model part, although this is highly unlikely. To provide an example, suppose there is a "FluidModelPart" that contains points that were generated from CAD and are not included in the computational model part.

File: applications/RomApplication/python_scripts/calculate_rom_basis_output_process.py (hunk 8 above)
Comment (SADPR) [reply applications/RomApplication/python_scripts/calculate_rom_basis_output_process.py]: As previously mentioned in my comment regarding this matter, I concur with Ruben's viewpoint. I further suggested that a future pull request should include the feature of allowing the user to pass the name of the "RomParameters" to the ``RomAnalysis`` function.

File: applications/RomApplication/tests/test_calculate_rom_basis_output_process.py (hunk 23 above)
Comment (Rbravo555) [reply applications/RomApplication/tests/test_calculate_rom_basis_output_process.py]: The issue here is, even when the field "rom_basis_output_name" seemed to provide the user the freedom to choose the name of the file to store the ROM information; in reality, the name RomPameters is hardcoded elsewhere within the workflow. Therefore, choosing a different name here, will invariably lead to a FILE_NOT_FOUND error.

In a future PR, we shall allow to name the file differently, which might come handy for "non-standard" simulations like coupled problems. Of course the solution shall involve other sections of the code and probably should happen within the RomManager #10763

File: applications/RomApplication/python_scripts/calculate_rom_basis_output_process.py [hunk 27]
```diff
@@ -151,20 +152,27 @@ def ExecuteFinalize(self):
         rom_basis_dict["rom_settings"]["number_of_rom_dofs"] = numpy.shape(u)[1] #TODO: This is way misleading. I'd call it number_of_basis_modes or number_of_rom_modes
         rom_basis_dict["projection_strategy"] = "galerkin" # Galerkin: (Phi.T@K@Phi dq= Phi.T@b), LSPG = (K@Phi dq= b), Petrov-Galerkin = (Psi.T@K@Phi dq = Psi.T@b)
```
Comment (SADPR) [inline applications/RomApplication/python_scripts/calculate_rom_basis_output_process.py:156]:
```suggestion
        rom_basis_dict["projection_strategy"] = "galerkin" # Galerkin: (Phi.T@K@Phi dq= Phi.T@b), LSPG = (K@Phi dq= b), Petrov-Galerkin = (Psi.T@K@Phi dq = Psi.T@b)
        rom_basis_dict["rom_format"] = self.rom_basis_output_format
```

File: applications/RomApplication/python_scripts/hrom_training_utility.py [hunk 28]
```diff
@@ -156,28 +159,12 @@ def __CalculateResidualBasis(self):
...
-                hrom_weights["Conditions"][int(z)-n_elements] = float(w)
-        else:
-            # Many elements found
-            for j in range (0,len(z)):
-                if z[j] <=  n_elements -1:
-                    hrom_weights["Elements"][int(z[j])] = float(w[j])
-                else:
-                    hrom_weights["Conditions"][int(z[j])-n_elements] = float(w[j])
+        # Create dictionary with HROM weights (Only used for the expansion of the selected Conditions to include their parent Elements)
+        hrom_weights = self.__CreateDictionaryWithRomElementsAndWeights(weights,indexes,number_of_elements)
```
Comment (SADPR) [inline applications/RomApplication/python_scripts/hrom_training_utility.py:171]: This function is called regardless of the format type ("json" or "numpy"). Inside this function, it tries to read numpy files that may not exist if the user is training with "json".

Comment (SADPR) [reply applications/RomApplication/python_scripts/hrom_training_utility.py:171]: Hyper-reduction training does not work for "json".

File: applications/RomApplication/python_scripts/hrom_training_utility.py (hunk 5 above)
Comment (SADPR) [reply applications/RomApplication/python_scripts/hrom_training_utility.py:214]: Yes, I think you are not rewriting the RomParameters file. I am not sure if this fits the new workflow, but I guess you are just missing the following lines:
```
with open('RomParameters.json','w') as f:
            json.dump(updated_rom_parameters, f, indent = 4)
```

Comment (Rbravo555) [reply applications/RomApplication/python_scripts/hrom_training_utility.py:214]: You are right. Its been added

File: applications/RomApplication/python_scripts/hrom_training_utility.py (hunk 28 above)
Comment (Rbravo555) [reply applications/RomApplication/python_scripts/hrom_training_utility.py:171]: Now it should work

File: applications/RomApplication/python_scripts/calculate_rom_basis_output_process.py (hunk 8 above)
Comment (Rbravo555) [reply applications/RomApplication/python_scripts/calculate_rom_basis_output_process.py]: The issue here is, even when the field "rom_basis_output_name" seemed to provide the user the freedom to choose the name of the file to store the ROM information; in reality, the name RomPameters is hardcoded elsewhere within the workflow. Therefore, choosing a different name here, will invariably lead to a FILE_NOT_FOUND error.

File: applications/RomApplication/python_scripts/calculate_rom_basis_output_process.py (hunk 26 above)
Comment (Rbravo555) [reply applications/RomApplication/python_scripts/calculate_rom_basis_output_process.py:169]: If that specific case does not work, it would not work for the JSON format either

File: applications/RomApplication/python_scripts/rom_analysis.py (hunk 18 above)
Comment (Rbravo555) [reply applications/RomApplication/python_scripts/rom_analysis.py]: A future PR should adress the naming of the RomParameters.json and all the numpy arrays.

**Comment by Rbravo555** [comment]
Since everything is working already, let us merge this PR and adapt #10763 accordingly

File: applications/RomApplication/tests/structural_static_test_files/HROM/RomParameters.json [hunk 29]
```diff
@@ -14,7 +15,10 @@
         ],
         "number_of_rom_dofs": 2
     },
-    "hrom_settings": {},
+    "hrom_settings": {
+        "hrom_format":"numpy",
+        "original_number_of_elements":116
```
Comment (SADPR) [inline applications/RomApplication/tests/structural_static_test_files/HROM/RomParameters.json]: Should this be here?

File: applications/RomApplication/tests/test_calculate_rom_basis_output_process.py [hunk 30]
```diff
@@ -33,7 +33,6 @@ def testCalculateRomBasisOutputProcess(self):
             "snapshots_interval": 1.0,
             "nodal_unknowns": ["TEMPERATURE"],
             "rom_basis_output_format": "json",
```
Comment (SADPR) [inline applications/RomApplication/tests/test_calculate_rom_basis_output_process.py]:
```suggestion
            "rom_basis_output_format": "json",
            "rom_basis_output_name": "RomParameters",
```

**Review by SADPR — COMMENTED** [review]
Two comments away from approval.

**Review by SADPR — APPROVED** [review]
All set. Ready.
